### ISSUE TITLE
Parallel creation of prefixes from pool result in duplicate error 409

### ISSUE THREAD

**Issue by kkaresz** [SpriteLink/NIPAP#1175]
When creating new prefixes from a pool with an automated process that sends bulks of create requests in parallel, some of the queries return error 409. The requests don't specify the IP address for the prefix, they only pass the from-pool argument expecting NIPAP to allocate the next available prefix. 

The reason for error 409 is that when 2 processes call the find_free_prefix() in nipap.backend it would return the same prefix for the 2 processes and only 1 of them would proceed successfully with the creation. It could only work if the database would be locked until the create request finishes either successfully or with a failure, so there could be no conflict.

Is this the intended behavior? If so, would a PR even be accepted that would aim to fix this problem?

I couldn't find any related open issues, but please link it here if it already exists.

**Comment by plajjan** [comment]
No, that is not the intended behaviour and I wonder if you are speculating on behaviour or how come you are saying this? What have you observed?

This was designed from the beginning to handle concurrency and we do that https://github.com/SpriteLink/NIPAP/blob/master/nipap/sql/functions.plsql#L474 by wrapping the whole prefix insertion in a loop. We try to insert and if that is successful we return, otherwise we continue spinning in the loop and try again. It is designed exactly to handle this sort of race.

So no, it doesn't require locking, we just retry until we are successful! :)

**Comment by kkaresz** [comment]
Thanks for the quick response and for the clarifications. You are right, I'm only speculating, sorry for not being well-prepared with proper debug details. I'm going to post as many as I can, let me check what am I doing wrong.

My observations that my ticket is based on is that when we call add_prefix with "from-pool" in the arguments we end up here: 
https://github.com/SpriteLink/NIPAP/blob/master/nipap/nipap/backend.py#L2645

which will look for the next free prefix here:
https://github.com/SpriteLink/NIPAP/blob/master/nipap/nipap/backend.py#L2985

and end up here:
https://github.com/SpriteLink/NIPAP/blob/master/nipap/sql/functions.plsql#L358

My theory was that if 2 processes call this at the same time, it is possible that both of them get the same result from the find_free_prefix function, and only one of them could successfully add the prefix here: 
https://github.com/SpriteLink/NIPAP/blob/master/nipap/nipap/backend.py#L2662

Again, please allow me to get proper stacktraces and debug logs, I should've collected them before opening this ticket.

The origin of the problem is that we try to add new prefixes in an automated fashion in a parallelized process (while creating new subnets in a virtual environment) and we occasionally get duplicate error 409 responses. We don't specify the prefix, only the pool name.

I'll get back with better details, I appreciate your time looking into this.

**Comment by garberg** [comment]
@kkaresz have you been able to look deeper into this?

**Comment by kkaresz** [comment]
@garberg I could not look deeper yet. Since we could work around the issue on our client side, the priorities shifted and I could not make time for it. I'll definitely update this thread once I managed to do some debugging. Also, sorry for the late response.
